File: arrays/n3-repeat-number.py
class Solution:
    def repeatedNumber(self, A):
        nby3 = len(A)/3
        a = 0; b = 0
        i = 0; j = 0

        for x in A:
            if x == a: i += 1
            elif x == b: j += 1
            elif i == 0: a = x; i = 1
            elif j == 0: b = x; j = 1
            else:
                i -= 1
                j -= 1

        i = 0; j = 0
        for x in A:
            if x == a: i += 1
            elif x == b: j+=1

        if i > nby3: return a
        if j > nby3: return b
        return -1

File: arrays/n3-repeat-number.cpp
int Solution::repeatedNumber(const vector<int> &A) {
    pair<int,int> a,b;
    int n = A.size();
    
    for(int i=0; i<n; i++){
        if(a.first == A[i]) a.second++;
        else if (b.first == A[i]) b.second++;
        else{
            if(a.second == 0){ a.first=A[i]; a.second++;}
            else if (b.second == 0){ b.first=A[i]; b.second++;}
            else{ a.second--; b.second--;}
        }
    }
    
    int ca=0, cb=0;
    for(int i=0; i<n; i++){
        if(A[i] == a.first) ca++;
        if(A[i] == b.first) cb++;
    }
 
    if(ca > n/3) return a.first;
    if(cb > n/3) return b.first;
    return -1;
}

